File: android/VelaWallet/app/src/main/java/app/getvela/wallet/service/LocalStorage.kt
package app.getvela.wallet.service

import android.content.Context
import android.content.SharedPreferences
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

private val json = Json { ignoreUnknownKeys = true }

class LocalStorage private constructor(private val prefs: SharedPreferences) {

    companion object {
        @Volatile
        private var instance: LocalStorage? = null

        fun init(context: Context) {
            if (instance == null) {
                synchronized(this) {
                    if (instance == null) {
                        instance = LocalStorage(
                            context.getSharedPreferences("vela_wallet", Context.MODE_PRIVATE)
                        )
                    }
                }
            }
        }

        val shared: LocalStorage
            get() = instance ?: throw IllegalStateException("LocalStorage not initialized. Call init() first.")
    }

    private val accountsKey = "vela.accounts"
    private val pendingUploadsKey = "vela.pendingUploads"
    private val customTokensKey = "vela.customTokens"

    // MARK: - Accounts

    @Serializable
    data class StoredAccount(
        val id: String,
        val name: String,
        val publicKeyHex: String,
        val address: String,
        val createdAt: Long = System.currentTimeMillis(),
    )

    fun saveAccount(account: StoredAccount) {
        val accounts = loadAccounts().toMutableList()
        accounts.removeAll { it.id == account.id }
        accounts.add(account)
        save(accountsKey, json.encodeToString(accounts))
    }

    fun loadAccounts(): List<StoredAccount> {
        val data = prefs.getString(accountsKey, null) ?: return emptyList()
        return try {
            json.decodeFromString(data)
        } catch (_: Exception) {
            emptyList()
        }
    }

    fun findAccount(credentialId: String): StoredAccount? =
        loadAccounts().find { it.id == credentialId }

    // MARK: - Pending Uploads

    @Serializable
    data class PendingUpload(
        val id: String,
        val name: String,
        val publicKeyHex: String,
        val attestationObjectHex: String,
        val createdAt: Long = System.currentTimeMillis(),
    )

    fun savePendingUpload(upload: PendingUpload) {
        val uploads = loadPendingUploads().toMutableList()
        uploads.removeAll { it.id == upload.id }
        uploads.add(upload)
        save(pendingUploadsKey, json.encodeToString(uploads))
    }

    fun loadPendingUploads(): List<PendingUpload> {
        val data = prefs.getString(pendingUploadsKey, null) ?: return emptyList()
        return try {
            json.decodeFromString(data)
        } catch (_: Exception) {
            emptyList()
        }
    }

    fun removePendingUpload(credentialId: String) {
        val uploads = loadPendingUploads().toMutableList()
        uploads.removeAll { it.id == credentialId }
        save(pendingUploadsKey, json.encodeToString(uploads))
    }

    fun hasPendingUploads(): Boolean = loadPendingUploads().isNotEmpty()

    // MARK: - Custom Tokens

    @Serializable
    data class CustomToken(
        val id: String,
        val chainId: Int,
        val contractAddress: String,
        val symbol: String,
        val name: String,
        val decimals: Int,
        val networkName: String,
    ) {
        val networkId: String
            get() = app.getvela.wallet.model.Network.networkId(chainId)
    }

    fun saveCustomToken(token: CustomToken) {
        val tokens = loadCustomTokens().toMutableList()
        tokens.removeAll { it.id == token.id }
        tokens.add(token)
        save(customTokensKey, json.encodeToString(tokens))
    }

    fun loadCustomTokens(): List<CustomToken> {
        val data = prefs.getString(customTokensKey, null) ?: return emptyList()
        return try {
            json.decodeFromString(data)
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun save(key: String, value: String) {
        prefs.edit().putString(key, value).apply()
    }
}

File: ios/VelaWallet/VelaWallet/Services/LocalStorage.swift
import Foundation

/// Persists wallet credential data to iCloud (NSUbiquitousKeyValueStore).
/// Syncs across devices automatically — switching phones won't lose data.
/// Falls back to UserDefaults if iCloud is unavailable.
final class LocalStorage {
    static let shared = LocalStorage()

    private let cloud = NSUbiquitousKeyValueStore.default
    private let local = UserDefaults.standard
    private let accountsKey = "vela.accounts"
    private let pendingUploadsKey = "vela.pendingUploads"
    private let customTokensKey = "vela.customTokens"

    private init() {
        // Trigger initial sync
        cloud.synchronize()

        // Migrate from UserDefaults to iCloud if needed
        migrateToCloudIfNeeded()
    }

    // MARK: - Accounts

    struct StoredAccount: Codable, Identifiable {
        let id: String           // credentialId (hex)
        let name: String
        let publicKeyHex: String // uncompressed P256: 04 || x || y
        let address: String      // Safe address
        let createdAt: Date
    }

    func saveAccount(_ account: StoredAccount) {
        var accounts = loadAccounts()
        accounts.removeAll { $0.id == account.id }
        accounts.append(account)
        save(accounts, forKey: accountsKey)
    }

    func loadAccounts() -> [StoredAccount] {
        load(forKey: accountsKey) ?? []
    }

    func findAccount(byCredentialId id: String) -> StoredAccount? {
        loadAccounts().first { $0.id == id }
    }

    // MARK: - Pending Uploads

    struct PendingUpload: Codable, Identifiable {
        let id: String           // credentialId
        let name: String
        let publicKeyHex: String
        let attestationObjectHex: String
        let createdAt: Date
    }

    func savePendingUpload(_ upload: PendingUpload) {
        var uploads = loadPendingUploads()
        uploads.removeAll { $0.id == upload.id }
        uploads.append(upload)
        save(uploads, forKey: pendingUploadsKey)
    }

    func loadPendingUploads() -> [PendingUpload] {
        load(forKey: pendingUploadsKey) ?? []
    }

    func removePendingUpload(credentialId: String) {
        var uploads = loadPendingUploads()
        uploads.removeAll { $0.id == credentialId }
        save(uploads, forKey: pendingUploadsKey)
    }

    func hasPendingUploads() -> Bool {
        !loadPendingUploads().isEmpty
    }

    // MARK: - Custom Tokens

    struct CustomToken: Codable, Identifiable {
        let id: String           // "{chainId}_{contractAddress}"
        let chainId: Int
        let contractAddress: String
        let symbol: String
        let name: String
        let decimals: Int
        let networkName: String  // display name like "Ethereum"

        var networkId: String {
            switch chainId {
            case 1: "eth-mainnet"
            case 42161: "arb-mainnet"
            case 8453: "base-mainnet"
            case 10: "opt-mainnet"
            case 137: "matic-mainnet"
            case 56: "bnb-mainnet"
            case 43114: "avax-mainnet"
            default: "eth-mainnet"
            }
        }
    }

    func saveCustomToken(_ token: CustomToken) {
        var tokens = loadCustomTokens()
        tokens.removeAll { $0.id == token.id }
        tokens.append(token)
        save(tokens, forKey: customTokensKey)
    }

    func loadCustomTokens() -> [CustomToken] {
        load(forKey: customTokensKey) ?? []
    }

    func removeCustomToken(id: String) {
        var tokens = loadCustomTokens()
        tokens.removeAll { $0.id == id }
        save(tokens, forKey: customTokensKey)
    }

    // MARK: - Private

    private func save<T: Encodable>(_ value: T, forKey key: String) {
        guard let data = try? JSONEncoder().encode(value) else { return }
        // Save to both iCloud and local
        cloud.set(data, forKey: key)
        cloud.synchronize()
        local.set(data, forKey: key)
    }

    private func load<T: Decodable>(forKey key: String) -> T? {
        // Prefer iCloud, fallback to local
        let data = cloud.data(forKey: key) ?? local.data(forKey: key)
        guard let data else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }

    private func migrateToCloudIfNeeded() {
        // If local has data but cloud doesn't, push to cloud
        for key in [accountsKey, pendingUploadsKey] {
            if cloud.data(forKey: key) == nil, let localData = local.data(forKey: key) {
                cloud.set(localData, forKey: key)
            }
        }
        cloud.synchronize()
    }
}
